### PR TITLE
Compact protocol generating structs - work in progress

### PR DESCRIPTION
**Pull request by paulanthonywilson**
PR to own master to help with getting feedback. Right now mainly just interested in getting serialisation and deserialisation of Structs working working.

## Approach:

* I've cloned `Thrift.Generator.BinaryProtocolTest` to `Thrift.Generator.CompactProtocolTest` and commented out the actual tests
* Created a `Thrift.Generator.StructCompactProtocol` and plugged it into the generator.
* Switched the `CompactProtocolTests` to use the compact serialisation. (Right now deserialisation is inactive.)

I've been using a combination of https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md as reference and Thrift's ruby binding to generate the test values.

One concern is that the 

I've been avoiding touching the existing code as much as possible: there's duplication.

Also the `CompactProtocol` module is being generated in the Structs whether you want it or not. Maybe this should be a configuration options.

Status:

* Highly work-in progress.
* Got the Bool test serialising (for fields, lists, maps, and sets).

Next:

1. Deserialising the bools
1. Draw the rest of the flipping owl.

### CODE REVIEW

**Comment by dantswain** [comment]
I wonder if there's a way to use the AST that's generated for the binary protocol and modify it to suit this?  I.e., generate the Binary protocol AST, change the serializer/deserializer functions, write out the modified AST.  That would avoid duplicating code, but could be more complex than is worthwhile?

**Comment by paulanthonywilson** [comment]
@dantswain I think that's probably beyond me, but apart from that I'm not so sure about it. It would be horrible to try and debug.

I suspect the duplication isn't as heavy as you might think; I'm going to have a shot at some deserialisation next and I think it might end up being very different as a lot of the fields that are fixed length in the binary protocol are now variable length.

**Comment by dantswain** [comment]
That's fair.  Thanks for giving this a shot!

File: lib/thrift/protocol/compact/container_item_type.ex [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  in struct fields. See https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md
+
+  Note that contrary to the above link, the type code for Bool seems to be 1, not 2,
+  in experiments with the Thrift Ruby implementation.
+  """
+
+  @typedoc "Compact protocol container item type identifiers"
+  @type t :: 1 | 3 | 4 | 6 | 8 | 10 | 11 | 12 | 13 | 14 | 15
+
+  defmacro bool,      do: 1
```
Comment (paulanthonywilson) [inline lib/thrift/protocol/compact/container_item_type.ex]: According to https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md#list-and-set bools are encoded as 2 within a map, list, or set. But the Ruby thrift implementation encodes them as 1.

**Comment by paulanthonywilson** [comment]
¯\\_(ツ)_/¯

**Pull request closed without merging**